sdk/go/docker/customTypes.go

```diff
@@ -0,0 +1,266 @@
+// Copyright 2016-2020, Pulumi Corporation.
+
+package docker
+
+import (
+	"context"
+	"reflect"
+
+	"github.com/pulumi/pulumi/sdk/go/pulumi"
+)
+
+type CacheFrom struct {
+	Stages []string `pulumi:"stages`
+}
+
+type CacheFromInput interface {
+	pulumi.Input
+
+	ToCacheFromOutput() CacheFromOutput
+	ToCacheFromOutputWithContext(context.Context) CacheFromOutput
+}
+
+type CacheFromArgs struct {
+	Stages pulumi.StringArrayInput `pulumi:"stages`
+}
+
+func (CacheFromArgs) ElementType() reflect.Type {
+	return reflect.TypeOf((*CacheFrom)(nil)).Elem()
+}
+
+func (i CacheFromArgs) ToCacheFromOutput() CacheFromOutput {
+	return i.ToCacheFromOutputWithContext(context.Background())
+}
+
+func (i CacheFromArgs) ToCacheFromOutputWithContext(ctx context.Context) CacheFromOutput {
+	return pulumi.ToOutputWithContext(ctx, i).(CacheFromOutput)
+}
+
+type CacheFromOutput struct{ *pulumi.OutputState }
+
+func (CacheFromOutput) ElementType() reflect.Type {
+	return reflect.TypeOf((*CacheFrom)(nil)).Elem()
+}
+
+func (o CacheFromOutput) ToCacheFromOutput() CacheFromOutput {
+	return o
+}
+
+func (o CacheFromOutput) ToCacheFromOutputWithContext(ctx context.Context) CacheFromOutput {
+	return o
+}
+
+type DockerBuild struct {
+	Context      string            `pulumi:"context"`
+	Dockerfile   string            `pulumi:"dockerfile"`
+	Args         map[string]string `pulumi:"args"`
+	CacheFrom    *CacheFrom        `pulumi:"cacheFrom"`
+	ExtraOptions []string          `pulumi:"extraOptions"`
+	Env          map[string]string `pulumi:"env"`
+	Target       string            `pulumi:"target"`
+}
+
+// DockerBuild may be used to specify detailed instructions about how to build a container.
```

Comment:
> Another thing I've noticed is that users sometimes get quite confused about which form of T TArgs TOutput TInput to use in various situations. 

+1 Because I was/am one of those users.

Reply:
Coming up with a standard comment here would be fantastic, yes.